lib/src/main/java/com/diffplug/spotless/scala/ScalaFixStep.java

```diff
@@ -0,0 +1,144 @@
+		Objects.requireNonNull(version, "version");
+		Objects.requireNonNull(scalaVersion, "scalaVersion");
+		Objects.requireNonNull(provisioner, "provisioner");
+		Objects.requireNonNull(scalaCompiler, "scalaCompiler");
+		return FormatterStep.createLazy(NAME, () -> new State(version, scalaVersion, provisioner, scalaCompiler, configFile), State::createFormat);
+	}
+
+	public static String defaultVersion() {
+		return DEFAULT_VERSION;
+	}
+
+	public static String defaultScalaVersion() {
+		return DEFAULT_SCALA_VERSION;
+	}
+
+	static final class State implements Serializable {
+		private static final long serialVersionUID = 1L;
+
+		final JarState jarState;
+		final FileSignature configSignature;
+		final ScalaCompiler scalaCompiler;
+
+		State(final String version, final String scalaVersion, final Provisioner provisioner, final ScalaCompiler scalaCompiler, @Nullable final File configFile) throws IOException {
+			this.jarState = JarState.from(String.format(MAVEN_COORDINATE, scalaVersion, version), provisioner);
+			this.configSignature = FileSignature.signAsList(configFile == null ? Collections.emptySet() : Collections.singleton(configFile));
+			this.scalaCompiler = scalaCompiler;
+		}
+
+		FormatterFunc createFormat() throws Exception {
+			final ClassLoader classLoader = jarState.getClassLoader();
+
+			final Class<?> javaConversionsCls = classLoader.loadClass("scala.collection.JavaConversions");
+			final Class<?> absolutePathCls = classLoader.loadClass("scala.meta.io.AbsolutePath");
+			final Class<?> classpathCls = classLoader.loadClass("scala.meta.io.Classpath");
+			final Class<?> someCls = classLoader.loadClass("scala.Some");
+			final Class<?> argsCls = classLoader.loadClass("scalafix.internal.v1.Args");
+			final Class<?> mainOpsCls = classLoader.loadClass("scalafix.internal.v1.MainOps");
+
+			final Method absolutePathApply = absolutePathCls.getMethod("apply", File.class, absolutePathCls);
+			final Method asScalaIterator = javaConversionsCls.getMethod("asScalaIterator", Iterator.class);
+			final Method argsDefault = argsCls.getMethod("default", absolutePathCls, PrintStream.class);
+			final Object workingDirectory = absolutePathCls.getMethod("workingDirectory").invoke(null);
+			final Object scalacOptionsIter = asScalaIterator.invoke(null, scalaCompiler.getCompilerOptions().iterator());
+			final Object scalacOptions = scalacOptionsIter.getClass().getMethod("toList").invoke(scalacOptionsIter);
+			final Object classpath = classpathCls.getMethod("apply", String.class).invoke(null, scalaCompiler.getClasspath());
+
+			final Method someApply = someCls.getMethod("apply", Object.class);
+
+			return new FormatterFunc() {
+				@Override
+				public String apply(String input) {
+					throw new RuntimeException("This method shouldn't be called.");
+				}
+
+				@Override
+				public String apply(String input, File source) throws Exception {
```

Comment:
Currently, we are ignoring `String input`.  The problem is that there might be other steps, say a license header step, which are called before this step, and this will wipe them out.  Also, we have an IDE integration with VSCode and others to come, which can format the dirty state of an editor.  So a user might type a bunch of code in a dirty buffer, and when they format this will wipe it out.

In the ideal world, we could pass the file to `scalafix` over `stdin` without tying it to the sourcetree on disk, but I assume that is not possible for the kinds of stuff that scalafix is doing.

So what we should do instead is:

1) read the current state of the `File source` to a `byte[]`, including its lastmodified timestamp (including handling the possibility that it is null or doesn't exist)
2) write `String input` to the file
  2a) you need the charset, which you don't have.  We can add it as a parameter to `FormatterFunc`, or we can assume UTF-8.  The right thing is probably the parameter, and we can do it in a backwards-compatible way.
  2b) before writing it out, I would read the file's content to check if there are any changes, since doing that one read will save you two writes
3) call the formatter
4) restore the state of the file to what it was like before (possibly deleting if necessary)

This hunk of code will be useful to other formatters that require to read the file from disk.  I would put it in `lib`, and name it something like `DiskBasedHelper`.

Reply:
I see. Thanks for your suggestion. Currently, the formatter depends on compilation of whole project which produces [semanticdb](https://scalameta.org/docs/semanticdb/guide.html) for each file, and Scalafix uses the metadata. If we can generate the metadata from a single `File source`, we can avoid the issue by following your idea. Let find a way to do that.

Reply:
So long as spotless depends on the compilation task, the IDE hook will still call the compile (or jar, whichever).  So it's fine to depend on compilation, so long as we can modify our one "being-formatted" file afterwards.  Very excited to have a PR of this sort, we'll definitely work with you to get it merged.